### PR TITLE
Create new arch module template: cpp-lib

### LINKED ISSUE #12480: Create a "new-lib" template, defaulting to C++, Fabric, Hermes, NuGet-based, Win32 lib

### PR DESCRIPTION
**Pull request by jonthysell**
## Description

Adds a new `cpp-lib` template designed to be applied on top of the new projects created by the (now recommended) `create-react-native-library` script. Also updates the `cpp-app` template to be able to autolink and run these new libraries.

### Type of Change
- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

### Why
Devs need to add windows support to native modules in a way which can be consumed by the new arch RNW apps (i.e. the `cpp-app` template).

Resolves #12480

### What
Create a new `cpp-lib` template which can be applied on top of the template created by `npx  create-react-native-library`, adding Windows support for a turbo module. It will also automatically apply the `cpp-app` template to the example project if it exists.

## Screenshots
image

## Testing
Verified the new template could be applied to a fresh library created by `create-react-native-library` and that example app autolinks, builds, and runs correctly.

## Changelog
Should this change be included in the release notes: yes

Created a new arch module template: cpp-lib
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/react-native-windows/pull/12481)

### CODE REVIEW

File: vnext/templates/cpp-lib/example/metro.config.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  transformer: {
+    getTransformOptions: async () => ({
+      transform: {
+        experimentalImportSupport: false,
+        inlineRequires: true,
+      },
+    }),
+
+    // This fixes the 'missing-asset-registry-path` error (see https://github.com/microsoft/react-native-windows/issues/11437)
+    assetRegistryPath: 'react-native/Libraries/Image/AssetRegistry',
```
Comment (acoates-ms) [inline vnext/templates/cpp-lib/example/metro.config.js:73]: Is this still needed?  We should try to keep this file as simple as possible.

File: vnext/templates/cpp-lib/example/metro.config.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+          `${path.resolve(__dirname, 'windows').replace(/[/\\]/g, '/')}.*`,
+        ),
+        // This prevents "react-native run-windows" from hitting: EBUSY: resource busy or locked, open msbuild.ProjectImports.zip or other files produced by msbuild
+        new RegExp(`${rnwPath}/build/.*`),
+        new RegExp(`${rnwPath}/target/.*`),
+        /.*\.ProjectImports\.zip/,
+      ])
+    ),
+
+    extraNodeModules: modules.reduce((acc, name) => {
```
Comment (acoates-ms) [inline vnext/templates/cpp-lib/example/metro.config.js:52]: Is this part of the default template? -- This is for monorepos, and we should align with what the main template is doing here.

File: vnext/templates/cpp-lib/windows/MyLib/pch.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+#define WIN32_LEAN_AND_MEAN 1
+#define WINRT_LEAN_AND_MEAN 1
+
+// Windows Header Files
+#include <windows.h>
+
+#pragma push_macro("GetCurrentTime")
+#undef GetCurrentTime
+
+// Playground pch.h
```
Comment (acoates-ms) [inline vnext/templates/cpp-lib/windows/MyLib/pch.h:20]: Probably shouldn't say playground.

File: vnext/templates/cpp-lib/example/metro.config.js (hunk 2 above)
Comment (jonthysell) [reply vnext/templates/cpp-lib/example/metro.config.js:52]: This file is what the `create-react-native-library` creates for it's example app - I just merged in the changes we make.

File: vnext/templates/cpp-lib/example/metro.config.js (hunk 1 above)
Comment (jonthysell) [reply vnext/templates/cpp-lib/example/metro.config.js]: Looks like it's not necessary anymore, taking it out.

**Review by acoates-ms — APPROVED** [review]

**Pull request merged**